Odata-docs/webapi-8/function-routing.md

```diff
@@ -0,0 +1,365 @@
+
+        employeeEntityType.Collection.Function("GetHighestRating")
+            .Returns<int>();
+        employeeEntityType.Function("GetRating")
+            .Returns<int>();
+        managerEntityType.Collection.Function("GetHighestBonus")
+            .Returns<decimal>();
+        managerEntityType.Function("GetBonus")
+            .Returns<decimal>();
+
+        services.AddControllers().AddOData(
+            options => options.EnableQueryFeatures(null).AddRouteComponents(
+                routePrefix: "odata",
+                model: modelBuilder.GetEdmModel()));
+    }
+
+    public void Configure(IApplicationBuilder app)
+    {
+        app.UseODataRouteDebug();
+        app.UseRouting();
+        app.UseEndpoints(endpoints => endpoints.MapControllers());
+    }
+}
+```
+
+---
+
+In the above block of code, we define an entity set named `Employees`. `Manager` is derived from `Employee` so it automatically gets included in the Edm model as an entity type.
+
+Four Edm functions are also defined:
+- `GetHighestRating` - bound to the `Employee` entity collection
+- `GetRating` - bound to the `Employee` entity
+- `GetHighestBonus` - bound to the `Manager` derived entity collection
+- `GetBonus` - bound to the `Manager` derived entity
+
+## Controller
+The partial structure of the controller for the OData service is as follows:
+
+```csharp
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.OData.Deltas;
+using Microsoft.AspNetCore.OData.Query;
+using Microsoft.AspNetCore.OData.Routing.Controllers;
+using FunctionRouting.Models;
+
+public class EmployeesController : ODataController
+{
+    private static IList<Employee> employees = new List<Employee>
+    {
+        new Employee { Id = 1, Name = "Employee 1", PerfRating = 8 },
+        new Employee { Id = 2, Name = "Employee 2", PerfRating = 7 },
+        new Employee { Id = 3, Name = "Employee 3", PerfRating = 5 },
+        new Employee { Id = 4, Name = "Employee 4", PerfRating = 3 },
+        new Manager { Id = 5, Name = "Employee 5", PerfRating = 7, Bonus = 2900 },
+        new Manager { Id = 6, Name = "Employee 6", PerfRating = 9, Bonus = 3700 }
+    };
+}
+```
+
+## Routing bound Edm functions
+In this section we cover the conventions for routing bound functions and the controller actions (endpoints) required for the requests to be routed successfully.
+
+### Invoking a function bound to an entity set or singleton
+The route template for this request is: `GET ~/{entityset}|{singleton}/{function}`
+
+The following request invokes the `GetHighestRating` function bound to the `Employees` entity set. The URL for the function is the function name appended to the entity set's URL:
+```http
+GET http://localhost:5000/odata/Employees/GetHighestRating()
+```
+
+An alternative URL for invoking the same function is: `http://localhost:5000/odata/Employees/Default.GetHighestRating()`.
```

Comment:
If we defined the model elements in the `FunctionRouting.Models` namespace.. why do we use `Default` here?

Reply:
@ElizabethOkerio In the section for building Edm model we don't change the Edm model namespace so it'll assume the value of `Default`. The structured elements always assume the namespace that the classes are defined in. Here's the $metadata for the actual service:
```xml
<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
    <edmx:DataServices>
        <Schema Namespace="FunctionRouting.Models" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <EntityType Name="Employee">
                <Key>
                    <PropertyRef Name="Id" />
                </Key>
                <Property Name="Id" Type="Edm.Int32" Nullable="false" />
                <Property Name="Name" Type="Edm.String" />
                <Property Name="PerfRating" Type="Edm.Int32" Nullable="false" />
            </EntityType>
            <EntityType Name="Manager" BaseType="FunctionRouting.Models.Employee">
                <Property Name="Bonus" Type="Edm.Decimal" Nullable="false" Scale="Variable" />
            </EntityType>
        </Schema>
        <Schema Namespace="Default" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <Function Name="GetHighestRating" IsBound="true">
                <Parameter Name="bindingParameter" Type="Collection(FunctionRouting.Models.Employee)" />
                <ReturnType Type="Edm.Int32" Nullable="false" />
            </Function>
            <Function Name="GetRating" IsBound="true">
                <Parameter Name="bindingParameter" Type="FunctionRouting.Models.Employee" />
                <ReturnType Type="Edm.Int32" Nullable="false" />
            </Function>
            <Function Name="GetHighestBonus" IsBound="true">
                <Parameter Name="bindingParameter" Type="Collection(FunctionRouting.Models.Manager)" />
                <ReturnType Type="Edm.Decimal" Nullable="false" Scale="Variable" />
            </Function>
            <Function Name="GetBonus" IsBound="true">
                <Parameter Name="bindingParameter" Type="FunctionRouting.Models.Manager" />
                <ReturnType Type="Edm.Decimal" Nullable="false" Scale="Variable" />
            </Function>
            <Function Name="GetSalary">
                <Parameter Name="hourlyRate" Type="Edm.Decimal" Nullable="false" Scale="Variable" />
                <Parameter Name="hoursWorked" Type="Edm.Int32" Nullable="false" />
                <ReturnType Type="Edm.Decimal" Nullable="false" Scale="Variable" />
            </Function>
            <EntityContainer Name="Container">
                <EntitySet Name="Employees" EntityType="FunctionRouting.Models.Employee" />
                <FunctionImport Name="GetSalary" Function="Default.GetSalary" IncludeInServiceDocument="true" />
            </EntityContainer>
        </Schema>
    </edmx:DataServices>
</edmx:Edmx>
```

There are two schema elements. The first one has the namespace `FunctionRouting.Models` while the second one has the namespace `Default` which is the, well, default. The actions and functions are under the `Default` namespace.